### PR TITLE
[Option 3] No typesafety at all

### PR DESCRIPTION
**Pull request by ryanpbrewster**
@leogrim @camhashemi @andrewconner

### CODE REVIEW

File: kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+}
+
+sealed trait Feature {
+  val value: String
+  def settings: Set[FeatureSetting]
+
+  private def toSetting(x: String): Option[FeatureSetting] = settings.find(_.value == x)
+  def settingReads: Reads[FeatureSetting] = __.read[String].map(x => toSetting(x).get)
+}
+sealed trait ExtraPermissions {
```
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala]: every time I see this name I have to ask "what does that mean again"? A better name could be "FeaturePermissions"

Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala]: It's a trait, so it gets mixed into specific features. Something like
`case object MyFavoriteFeature extends Feature with ExtraPermissions`

Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala]: sure, but still not sure what the trait specifies

i.e. when does a `Feature extends ExtraPermissions` and when does it not

**Comment by leogrim** [comment]
lgtm, let's run with this

File: kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  def withSettings(newKvs: (Feature, FeatureSetting)*): OrganizationSettings = setAll(newKvs.toMap)
+  def extraPermissionsFor(roleOpt: Option[OrganizationRole]): Set[OrganizationPermission] = kvs.collect {
+    case (feature: FeatureWithPermissions, setting: FeatureSetting) => feature.extraPermissionsFor(roleOpt, setting)
+  }.toSet.flatten
+
+  def diff(that: OrganizationSettings): Set[Feature] = {
+    this.kvs.collect {
+      case (feature, setting) if that.kvs(feature) != setting => feature
+    }.toSet
+  }
```
Comment (camhashemi) [inline kifi-backend/modules/common/app/com/keepit/model/OrganizationSettings.scala:18]: noice

**Comment by ryanpbrewster** [comment]
Okay team, I'm gonna merge this in about an hour. I'm hoping everything will go smoothly, but I have a healthy amount of skepticism.

**Comment by andrewconner** [comment]
Looks good to me. :+1: merge away

File: kifi-backend/modules/common/conf/evolutions/shoebox/398.sql [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ALTER TABLE paid_plan ADD COLUMN editable_features text NOT NULL;
+ALTER TABLE paid_plan ADD COLUMN default_settings text NOT NULL;
+
+CREATE TABLE organization_configuration (
+    id bigint(20) NOT NULL AUTO_INCREMENT,
+    created_at datetime NOT NULL,
+    updated_at datetime NOT NULL,
+    state varchar(20) NOT NULL,
+    organization_id bigint(20) NOT NULL,
+    settings text NOT NULL,
```
Comment (camhashemi) [inline kifi-backend/modules/common/conf/evolutions/shoebox/398.sql:19]: could be nice at some point to have a column for each feature, then we can just add/drop/edit a column whenever we want.

File: kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentRequest.scala [hunk 4]
```diff
@@ -1,7 +1,7 @@
 package com.keepit.payments
 
-import com.keepit.common.db.ExternalId
-import com.keepit.model.{ Name, User }
+import com.keepit.common.db.{ Id, ExternalId }
+import com.keepit.model._
```
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/payments/PaymentRequest.scala:4]: :(

**Comment by camhashemi** [comment]
:+1: lgtm.

**Pull request merged**